### PR TITLE
fix: Don't unify arrays with their elements

### PR DESCRIPTION
**Pull request by lnicola**
Addresses https://github.com/rust-analyzer/rust-analyzer/pull/9526#issuecomment-876206100.

### CODE REVIEW

File: crates/hir_ty/src/infer/expr.rs [hunk 1]
```diff
@@ -738,7 +738,8 @@ impl<'a> InferenceContext<'a> {
                 let len = match array {
                     Array::ElementList(items) => {
                         for expr in items.iter() {
-                            let cur_elem_ty = self.infer_expr_inner(*expr, expected);
+                            let cur_elem_ty = self
+                                .infer_expr_inner(*expr, &Expectation::has_type(elem_ty.clone()));
```
Comment (flodiebold) [inline crates/hir_ty/src/infer/expr.rs:742]: I think we should only use the original expectation, not the modified `elem_ty`

**Comment by flodiebold** [comment]
We should also add a test for this (one similar to the other one with `check_no_mismatches` but with an explicit type on the `let` should work?), if you don't mind?

File: crates/hir_ty/src/infer/expr.rs (hunk 1 above)
Comment (lnicola) [reply crates/hir_ty/src/infer/expr.rs:742]: Do you mean like this?

```diff
diff --git i/crates/hir_ty/src/infer/expr.rs w/crates/hir_ty/src/infer/expr.rs
index 71a1088b4..fea3264ec 100644
--- i/crates/hir_ty/src/infer/expr.rs
+++ w/crates/hir_ty/src/infer/expr.rs
@@ -734,12 +734,11 @@ impl<'a> InferenceContext<'a> {
                         Some(TyKind::Array(st, _) | TyKind::Slice(st)) => st.clone(),
                         _ => self.table.new_type_var(),
                     };
-
+                let expected = Expectation::has_type(elem_ty.clone());
                 let len = match array {
                     Array::ElementList(items) => {
                         for expr in items.iter() {
-                            let cur_elem_ty = self
-                                .infer_expr_inner(*expr, &Expectation::has_type(elem_ty.clone()));
+                            let cur_elem_ty = self.infer_expr_inner(*expr, &expected);
                             elem_ty = self.coerce_merge_branch(Some(*expr), &elem_ty, &cur_elem_ty);
                         }
                         Some(items.len() as u64)
```

Comment (flodiebold) [reply crates/hir_ty/src/infer/expr.rs:742]: exactly

**Comment by lnicola** [comment]
I tried this as a test:

```rust
#[test]
fn recursive_vars_3() {
    check_infer(
        r#"
        fn test() {
            let x: i32 = unknown;
            let y = unknown;
            [(x, y), (y, x)];
        }
        "#,
        expect![[r#"
            10..82 '{     ...x)]; }': ()
            20..21 'x': i32
            29..36 'unknown': i32
            46..47 'y': i32
            50..57 'unknown': i32
            63..79 '[(x, y...y, x)]': [(i32, i32); 2]
            64..70 '(x, y)': (i32, i32)
            65..66 'x': i32
            68..69 'y': i32
            72..78 '(y, x)': (i32, i32)
            73..74 'y': i32
            76..77 'x': i32
        "#]],
    );
}
```

But it works both in `master` and in my first attempt at this.

**Comment by flodiebold** [comment]
I was more thinking something simple like
```rust
#[test]
fn array_elements_expected_type() {
    check_types(
        r#"
fn test() {
    let x: [u32; 2] = [1, 2];
                       // ^ u32
}
        "#,
    );
}
```
Does that work without this PR?

**Comment by lnicola** [comment]
Yeah, it passes.

**Comment by flodiebold** [comment]
Ah right, because 1. the correct type gets inferred through the unification of the array type later, and 2. we don't record type mismatches for the element types from the expected type. How about
```rust
#[test]
fn array_elements_expected_type() {
    check_no_mismatches(
        r#"
fn test() {
    let x: [[u32; 2]; 2] = [[1, 2], [3, 4]];
}
        "#,
    );
}
```
?

**Comment by lnicola** [comment]
That works, thank you.

**Comment by flodiebold** [comment]
> Do we also want a regression test for my original attempt?

If you have an idea how to do it sure, but it might require some thinking :thinking:

**Comment by lnicola** [comment]
Nope, not today :sweat_smile:.

**Review by flodiebold — APPROVED** [review]
r=me

**Comment by lnicola** [comment]
bors r=flodiebold

(CI error is spurious)

**Comment by lnicola** [comment]
bors retry

**Pull request merged**